tests/verify.rs

```diff
@@ -1,8 +1,170 @@
+    // that contains a DNS name matching the permitted subtree base.
+    let permitted_certs = certs_for_subtrees(
+        &trust_anchor,
+        name_constraints.permitted_subtrees.iter(),
+        |dns_name| format!("valid{}", dns_name).to_string(),
+    );
+
+    // For each permitted subtree in the name constraints, issue an end entity certificate
+    // that contains a DNS name that will **not** match the permitted subtree base.
+    let forbidden_certs = certs_for_subtrees(
+        &trust_anchor,
+        name_constraints.permitted_subtrees.iter(),
+        |dns_name| format!("valid{}.invalid", dns_name).to_string(),
+    );
+
+    ConstraintTest {
+        trust_anchor: trust_anchor.serialize_der().unwrap(),
+        permitted_certs,
+        forbidden_certs,
+    }
+}
+
+fn rcgen_ee_for_name(name: String, issuer: &Certificate) -> Vec<u8> {
+    let mut ee = CertificateParams::new(vec![name.clone()]);
+    ee.distinguished_name.push(DnType::CommonName, name);
+    ee.is_ca = IsCa::NoCa;
+    Certificate::from_params(ee)
+        .unwrap()
+        .serialize_der_with_signer(&issuer)
+        .unwrap()
+}
+
+/// Convert the webpki trust anchor DER encoding of name constraints to rcgen NameConstraints.
+fn rcgen_name_constraints(der: &[u8]) -> NameConstraints {
+    // x509 parser expects the outer SEQUENCE that the webpki trust anchor representation elides
+    // so wrap the DER up.
+    //
+    // Note: We take the cheap way out here and assume single byte length - if the following
+    // assert fails we'll need to more intelligently encode the sequence DER length.
+    assert!(der.len() < 0x80, "name constraint too long");
+    let wrapped_der = [&[0x30, der.len() as u8], der].concat();
+
+    // Constraints should parse with no trailing data.
+    let (trailing, constraints) = X509ParserNameConstraints::from_der(&wrapped_der).unwrap();
+    assert!(
+        trailing.is_empty(),
+        "unexpected trailing DER in name constraint"
+    );
+
+    // There should be at least one permitted or excluded subtree.
+    assert!(
+        !constraints.permitted_subtrees.is_none(),
+        "empty permitted subtrees in constraints"
+    );
+
+    // We don't expect any excluded subtrees as this time.
+    assert!(constraints.excluded_subtrees.is_none());
+
+    // Collect all of the DNS names from the x509-parser representation, mapping to the rcgen
+    // representation usable in cert parameters. We don't expect to find any other types of general
+    // name and x509-parser doesn't parse the subtree minimum and maximum (which we would assert to
+    // be missing for proper encoding anyway).
+    let permitted_subtrees = match constraints.permitted_subtrees {
+        None => Vec::default(),
+        Some(subtrees) => subtrees
+            .iter()
+            .map(|subtree| match &subtree.base {
+                GeneralName::DNSName(base) => GeneralSubtree::DnsName(base.to_string()),
```

Comment:
Here I would probably also keep the `rcgen::` prefix for `GeneralSubtree`.

Reply:
Done.